.github/workflows/linting.yml

```diff
@@ -0,0 +1,80 @@
+---
+name: Linting
+
+on:
+  pull_request:
+    branches:
+      - '**'
+  push:
+    branches:
+      - master
+    tags:
+      - 'v**'
```

Comment:
Is there a benefit to run these on push even?
They already run on each PR, so most of them should be fine on push.

Perhaps it only makes sense to run those that could've changed if the PR was sitting a whjile before push (like the MD links since that's affected by 3rd party sources)

Reply:
If I understand this triggering correctly, it's only running when pushing tags. To me it makes sense to run absolutely everything when we're tagging.

Reply:
I marked lines 6-10 which are run on every push not just tags

Except perhaps MD links (which is 3rd party effected) I don't think there's a benefit to re-run all linting on the push as it should've already passed in the PR

Even for MD links there's limited value to run on push since if the job actually fails you won't be able to change the commit and you'll have to fix it in a PR anyhow.

Reply:
Is there a situation where we push something other than a tag? I guess we don't have automation to prevent that, but maybe we should. I thought all commits should go through PRs, never direct pushes.

Reply:
I guess https://docs.github.com/en/github/administering-a-repository/about-branch-restrictions for master and https://docs.github.com/en/github/administering-a-repository/enabling-required-status-checks for all checks, with "Include admins", would get us part-way there. As I understand, pushes would be rejected to master if all required checks don't pass.

We could also add a job that raises an Issue to report direct pushes, like the example here: https://github.com/marketplace/actions/create-an-issue#usage.

Reply:
I think you missed how GHA works, for every PR that's merged a "push" event is created which runs the tests you specify in the file that is listening to the push.
If the push was specifying only to listen to tags then it would only trigger if a tag was pushed, but as it's written here it'll trigger for any push to `master` (i.e. every PR merged to master) and for any tag starting with `v`.

For example you can see here the merge appearing as a push event:
![image](https://user-images.githubusercontent.com/51078027/93198813-3401c200-f756-11ea-8463-a621c55a5d8c.png)

Reply:
So you're saying it's running twice on the same code for normal PRs - once for the PR check and then again when the PR is actually merged? It seems like that would be mentioned in https://docs.github.com/en/actions/reference/events-that-trigger-workflows#push.

I understand that this will trigger the jobs on both pushes to master and PRs, and that's what I'm saying is desirable above. The jobs triggering on pushes is required for us to enforce these checks pass for direct pushes to master for all users.

Reply:
Even if they jobs re-run when PRs are merged, is that a bad thing? Don't we want tests job that directly parallel release jobs, not just verify PRs?

Reply:
For example, one could imagine that a PR could pass checks that the same code, after "Rebase and Merge" on top of a logically-related change, would fail. We want a job to catch that imho.

Reply:
> So you're saying it's running twice on the same code for normal PRs - once for the PR check and then again when the PR is actually merged?

Yes at least twice (it's also re run when you update the PR)

> 
> I understand that this will trigger the jobs on both commit/tag pushes to master and PRs, and that's what I'm saying is desirable above. The jobs triggering on pushes is required for us to enforce these checks pass for direct pushes to master for all users.

AFAIK direct pushes are protected by GH policy so this is not a concern.

> Even if the jobs re-run when PRs are merged, is that a bad thing? Don't we want test jobs that directly parallel release jobs, not just verify PRs?

Not sure what you mean, can you please clarify?

> For example, one could imagine that a PR could pass checks that the same code, after "Rebase and Merge" on top of a logically-related change, would fail. We want a job to catch that imho.

While for unit tests and E2E tests that would be a desirable action since even if the code "merges nicely" it could still break something after rebase, I don't think the same can be said of linting.
I don't see how suddenly a DCO check would fail or MarkDown would stop linting after such a scenario, if the rebase was clean.
I guess the only exception is if you change the linting behavior itself and merge that before the patch, which would be quite the corner case and I dont think makes sense to run these checks on merge just for that.

Originally many of these checks didnt use to run on merge, and now they would, this needs to be justified since every job run we add takes time and resources which couldve been spent on other jobs (not to mention downloading docker images needlessly) and hence I raised this point about the linting.

One more thing to consider - these jobs (that run after push) aren't gating so if they fail the merge won't be affected (as it has already happened), so it's really up to an observant maintainer to notice if they even fail.

Reply:
> AFAIK direct pushes are protected by GH policy so this is not a concern.

Right, but only via the branch protection and required status checks I've been talking about. So we have:

1. Branch protection for master, so only admins can directly push.
2. Required status checks for master, so admins can only directly push if required checks pass.

So to prevent admins from pushing commits that don't pass required checks, those checks have to create statuses on pushes.

> One more thing to consider - these jobs (that run after push) aren't gating so if they fail the merge won't be affected (as it has already happened), so it's really up to an observant maintainer to notice if they even fail.

That's not true with the configuration I'm describing:

> You won't be able to push local changes to a protected branch until all required status checks pass. Instead, you'll receive an error message similar to the following:
remote: error: GH006: Protected branch update failed for refs/heads/main.
remote: error: Required status check "ci-build" is failing

https://docs.github.com/en/github/administering-a-repository/about-required-status-checks#troubleshooting-required-status-checks

Reply:
We talked about this quite a bit on the automation sync just now, summarizing:

* Required status checks aren't reported for push events, only PRs, which invalidates much of what I was thinking above.
* There is config for requiring branches (and so also statues) to be up-to-date before merging, which does most/all of the work that I was looking for.
* With the config now on this repo (should be copied to other repos), I think all my concerns about statues applying to admins are addressed.
* We want to cause all jobs to run on tags, but only release jobs to run on pushes to master.

Reply:
Okay, I think it's fixed in the latest. Thanks @mkolesnik!